r/tests/testthat/test-dplyr.R

```diff
@@ -369,3 +384,46 @@ test_that("tail", {
       group_by(int)
     )
 })
+
+test_that("relocate", {
+  df <- tibble(a = 1, b = 1, c = 1, d = "a", e = "a", f = "a")
+  expect_dplyr_equal(
+    input %>% relocate(f) %>% collect(),
+    df,
+  )
+  expect_dplyr_equal(
+    input %>% relocate(a, .after = c) %>% collect(),
+    df,
+  )
+  expect_dplyr_equal(
+    input %>% relocate(f, .before = b) %>% collect(),
+    df,
+  )
+  expect_dplyr_equal(
+    input %>% relocate(a, .after = last_col()) %>% collect(),
+    df,
+  )
+  expect_dplyr_equal(
+    input %>% relocate(ff = f) %>% collect(),
+    df,
+  )
+})
+
+test_that("relocate with selection helpers", {
+  expect_dplyr_equal(
+    input %>% relocate(any_of(c("a", "e", "i", "o", "u"))) %>% collect(),
+    df
+  )
+  expect_error(
+    df %>% Table$create() %>% relocate(where(is.character)),
```

Comment:
What is the failure mode without `check_select_helpers()`?

Likewise with select/rename: how do they fail differently without checking? (If we're adding the check, we should add tests too)

How does dbplyr et al. handle these? Surely you would have the same problem of `select(where(is.factor))` any time you're dealing with a non-data.frame backend.

Reply:
Nope, this stuff fails in dbplyr 
```r
con <- DBI::dbConnect(RSQLite::SQLite(), path = ":dbname:")
copy_to(con, nycflights13::flights, "flights")
flights_db <- tbl(con, "flights")
flights_db %>% select(where(is.numeric))
## Error: Query contains no columns
```

Reply:
Without `check_select_helpers()`, `relocate()` fails silently:
```r
df <- tibble(a = 1, b = 1, c = 1, d = "a", e = "a", f = "a")
df %>% Table$create() %>% relocate(where(is.numeric), .after = f)
## Table (query)
## a: double
## b: double
## c: double
## d: string
## e: string
## f: string
## 
## See $.data for the source Arrow object
```

And here's how `select()` and `rename()` fail:
```r
df <- tibble(a = 1, b = 1, c = 1, d = "a", e = "a", f = "a")
df %>% Table$create() %>% select(where(is.numeric))
##  Error: This tidyselect interface doesn't support predicates yet.
## ℹ Contact the package author and suggest using `eval_select()`.
## Run `rlang::last_error()` to see where the error occurred. 
```
I believe this error happens only because we're using questioning-stage rlang functions (`vars_select()` and `vars_rename()`) in our `column_select()` function and they don't support `where()`.

Reply:
I'm going to keep looking to see if there's anything in rlang or dplyr or vctrs that could obviate the need for us to perform this check in arrow code.

Reply:
I can't find any prefab solution to this. I also tried creating an environment in which `where` is defined as `stop()` and passing that environment to `eval_select()`, but that doesn't solve it.

Reply:
Re vars_select, it appears it has been deprecated: https://github.com/r-lib/tidyselect/blob/fe65d19ccdc5f5ead4d4d3c8d869926f6dccec3b/NEWS.md#L214

Reply:
Huh, the docs says they are "questioning" but not yet deprecated https://github.com/r-lib/tidyselect/blob/69fdc96fd9f47a7ae5563c33d356ae67dc0799e8/R/lifecycle-deprecated.R#L5-L7
Either way, we should try to replace them with `eval_select()`. I made an attempt to do that as part of this PR but it got very complicated and I backed out. I'll open a Jira for that.

Reply:
SGTM, it seems like the advantage of eval_select is that you can evaluate the kinds of predicates that we're trying to prevent evaluating anyway. (Aside: I wonder if switching `relocate` to use `vars_select` would mean that you didn't have to bother (for now) filtering out `where` et al. because vars_select would error for you)

Reply:
ARROW-12105

Reply:
I don't like the idea of switching this to use `vars_select()` since we really ought to switch everything away from them as soon as possible

Reply:
For reference, here's what dbplyr does: https://github.com/tidyverse/dbplyr/blob/master/R/verb-select.R